aten/src/ATen/native/Indexing.cpp

```diff
@@ -314,6 +314,46 @@ Tensor index_copy(const Tensor & self, int64_t dim, const Tensor & index, const
   return self.clone().index_copy_(dim, index, source);
 }
 
+
+Tensor& index_add__cpu(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  dim = maybe_wrap_dim(dim, self.dim());
+
+  auto numel = index.numel();
+  TORCH_CHECK_INDEX(index.dim() <= 1, "index_add_(): Index is supposed to be a vector");
+  TORCH_CHECK(index.scalar_type() == ScalarType::Long, "index_add_(): Expected LongTensor for index");
+  TORCH_CHECK(self.scalar_type() == source.scalar_type(),
+              "index_add_(): self and source must have the same scalar type");
+
+  auto index_contig = index.contiguous();
+  auto index_data = index_contig.data_ptr<int64_t>();
+
+  if (self.dim() > 1) {
+    for (auto i=0; i<numel; i++) {
+      auto selfSlice = self.select(dim, index_data[i]);
+      auto sourceSlice = source.select(dim, i);
+      selfSlice.add_(sourceSlice, 1);
+    }
+  }
+  else {
+    TORCH_CHECK_INDEX(source.dim() <= 1, "source tensor must have one dimension");
```

Comment:
nit: I'd just fix these error messages, i.e. here say "source tensor must have 0 or 1 dimensions".

Reply:
I've changed the error messages, although I'm not sure this is the right direction.
Documentation for `index_add_` states `index` must be a vector., and talks about the `dim`th dimension of `tensor`, so this support of scalar `index` and `tensor` contradicts the documentation. Also the operator isn't very useful when `tensor` / `index` is a scalar. I suspect the support for scalars is mostly an artifact of TH api.

I would prefer deprecating scalars for `index` and `source` it instead of "documenting" it in the error messages.